File: Algo/Leetcode/438FindAllAnagramsInAString.cpp
// Use a vector to comp each sliding window
// Time is O(m + (n-m) * 26), Space is O(26)
class Solution {
public:
    bool checkAllZero(vector<int> count){
        for(int i:count)
            if(i!=0) return false;
        return true;
    }

    vector<int> findAnagrams(string s, string p){
        vector<int> res;
        int n = s.length();
        int m = p.length();
        if(m > n) return res;

        vector<int> count(26, 0);

        for(int i=0; i<m; i++){
            count[p[i]-'a'] -= 1;
            count[s[i]-'a'] -= 1;
        }

        if(checkAllZero(count)) res.push_back(0);

        for(int i=m; i<n; i++){
            count[s[i]-'a'] -= 1;
            count[s[i-m]-'a'] += 1;
            if(checkAllZero(count))
                res.push_back(i-m+1);
        }

        return res;
    }
};


// You do not need to comp the whole array each pos, but use a single diff value to count
// The value is changed based on the coming change
// Time is O(m + 26 + (n-m)), Space is O(26)
class Solution {
public:
    int countDiff(vector<int> count){
        int diff = 0;
        for(int i:count){
            if(i!=0)
                diff += 1;
        }
        return diff;
    }

    vector<int> findAnagrams(string s, string p){
        vector<int> res;
        int n = s.length();
        int m = p.length();
        if(m > n) return res;

        vector<int> count(26, 0);
        for(int i=0; i<m; i++){
            count[p[i]-'a'] -= 1;
            count[s[i]-'a'] -= 1;
        }
        int diff = countDiff(count);

        if(diff==0) res.push_back(0);

        for(int i=m; i<n; i++){
            if(count[s[i-m]-'a']==-1)       // the leaving one will lead to a reduced diff
                diff -= 1;
            else if(count[s[i-m]-'a']==0)   // the leaving one will lead to a increased diff
                diff += 1;
            count[s[i-m]-'a'] += 1;

            if(count[s[i]-'a'] == 1)        // the coming one will lead to a reduced diff
                diff -= 1;
            else if(count[s[i]-'a']==0)     // the coming one will lead to a increase diff
                diff += 1;
            count[s[i]-'a'] -= 1;

            if(diff == 0)
                res.push_back(i-m+1);
        }

        return res;
    }
};

File: Algo/Leetcode/438FindAllAnagramsInAString.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        res = []
        n = len(s)
        m = len(p)
        if n < m:  # s < p always
            return res

        table = [0] * 26
        for i in range(m):  # init the table
            table[getOrd(p[i])] -= 1
            table[getOrd(s[i])] += 1
        diff = countDiff(table)
        if diff == 0:
            res.append(0)

        for i in range(m, n):  # check in longer p
            if table[getOrd(s[i - m])] == 0:
                diff += 1
            elif table[getOrd(s[i - m])] == 1:
                diff -= 1
            table[getOrd(s[i - m])] -= 1

            if table[getOrd(s[i])] == 0:
                diff += 1
            elif table[getOrd(s[i])] == -1:
                diff -= 1
            table[getOrd(s[i])] += 1

            if diff == 0:
                res.append(i - m + 1)

        return res


def countDiff(table):
    count = 0
    for n in table:
        if n != 0:
            count += 1

    return count


def getOrd(c):
    return ord(c) - ord('a')